File: solutions/java/src/digitalwallet/Account.java
package digitalwallet;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

public class Account {
    private final String id;
    private final User user;
    private final String accountNumber;
    private final Currency currency;
    private BigDecimal balance;
    private final List<Transaction> transactions;

    public Account(String id, User user, String accountNumber, Currency currency) {
        this.id = id;
        this.user = user;
        this.accountNumber = accountNumber;
        this.currency = currency;
        this.balance = BigDecimal.ZERO;
        this.transactions = new ArrayList<>();
    }

    public synchronized void deposit(BigDecimal amount) {
        balance = balance.add(amount);
    }

    public synchronized void withdraw(BigDecimal amount) {
        if (balance.compareTo(amount) >= 0) {
            balance = balance.subtract(amount);
        } else {
            throw new InsufficientFundsException("Insufficient funds in the account.");
        }
    }

    public synchronized void addTransaction(Transaction transaction) {
        transactions.add(transaction);
    }

    public String getId() {
        return id;
    }

    public User getUser() {
        return user;
    }

    public String getAccountNumber() {
        return accountNumber;
    }

    public Currency getCurrency() {
        return currency;
    }

    public BigDecimal getBalance() {
        return balance;
    }

    public List<Transaction> getTransactions() {
        return transactions;
    }
}

File: solutions/golang/digitalwalletservice/account.go
package digitalwallet

import (
	"math/big"
	"sync"
)

type Account struct {
	ID            string
	User          *User
	AccountNumber string
	Currency      Currency
	balance       *big.Float
	transactions  []*Transaction
	mu            sync.RWMutex
}

func NewAccount(id string, user *User, accountNumber string, currency Currency) *Account {
	return &Account{
		ID:            id,
		User:          user,
		AccountNumber: accountNumber,
		Currency:      currency,
		balance:       big.NewFloat(0),
		transactions:  make([]*Transaction, 0),
	}
}

func (a *Account) Deposit(amount *big.Float) {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.balance.Add(a.balance, amount)
}

func (a *Account) Withdraw(amount *big.Float) error {
	a.mu.Lock()
	defer a.mu.Unlock()

	if a.balance.Cmp(amount) >= 0 {
		a.balance.Sub(a.balance, amount)
		return nil
	}
	return NewInsufficientFundsError("Insufficient funds in the account")
}

func (a *Account) AddTransaction(transaction *Transaction) {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.transactions = append(a.transactions, transaction)
}

func (a *Account) GetBalance() *big.Float {
	a.mu.RLock()
	defer a.mu.RUnlock()
	return new(big.Float).Copy(a.balance)
}

func (a *Account) GetTransactions() []*Transaction {
	a.mu.RLock()
	defer a.mu.RUnlock()
	return append([]*Transaction{}, a.transactions...)
}
